Dockerfile-tests

```diff
@@ -0,0 +1,46 @@
+# We build an image with bee-queue's node_modules and all the source
+# files.  We jump through some Docker FROM hoops to avoid having the
+# bloat of the g++ compiler and other tools in the final image.
+
+FROM node:lts-alpine as beequeue-base
+
+# Info, and add bash, which we need in the final image
+RUN true \
+  && echo -n "node " && node --version \
+  && echo -n "npm " && npm --version \
+  && echo -n "apk " && apk --version \
+  && apk update \
+  && apk add --no-cache bash \
+  && rm -r /usr/share/man \
+  && true
+
+FROM beequeue-base as beequeue-node_modules
+
+# Ensure that tools for npm are available (for node-gyp in particular)
+# to build node_modules
+RUN true \
+  && apk add --no-cache python make g++ \
+  && true
+
+WORKDIR /bee-queue
+
+# Start with just these two files so as to optimize that npm install
+# only runs when package.json has been updated
+COPY package.json package-lock.json ./
+RUN npm --version && npm ci
+
+# Build the target image
+FROM beequeue-base
+
+WORKDIR /bee-queue
+
+COPY --from=beequeue-node_modules /bee-queue/node_modules ./node_modules/
+
+# We use a script for Docker-based testing (instead of npm test) so
+# that the dev cycle doesn't have to wait for the npm install
+# triggered by editing e.g. the scripts stanza in package.json
```

Comment:
I'm not sure I follow this - if I'm not mistaken, `npm test` just executes the commands defined in the `test` script - how does `npm install` factor into it? I'd like to avoid duplicating the commands defined in the `package.json` if at all possible.

Reply:
We're trying to avoid having `npm ci` get run each time we tweak a test.  Docker is clever about caching and not redoing work it can prove it already has the results of, but it doesn't parse files for semantics.  So, if `package.json` is altered in any way from prior builds, the `npm ci` step gets re-run.  By relying on a separate script, that's copied into the image file system after the `npm ci` step,  that step is not re-run and the dev cycle is much quicker.  Having a script also supports running a custom script and other dev-type work -- nothing says you have to run tests in the script....

I'm approaching this the way I work in other projects -- I'm new to the Bee-Queue culture and if/how you want to enable container-based development and use of Bee-Queue.

Reply:
Thanks for sharing - it's interesting to hear how folks use Docker in development scenarios. I've only really used it for shipping production code in deployment scenarios. I'm cool with this, though I feel like the tweaking of the `scripts` object is rare enough that this is somewhat of a weird workflow optimization.